### DISCUSSION TITLE
NextApiRequest vs NextRequest headers

### DISCUSSION THREAD

**Discussion by tgerakitis** [vercel/next.js#38596] [Help]
Hello,

I'm new to next.js and have a question about parsing headers in different parts of the code.

I want to parse the header injected by my reverse proxy and read specific data from it in different places.

There are two big differences that I have encountered so far:
When coming from the middleware, `req ` is a `NextRequest` object. (https://nextjs.org/docs/advanced-features/middleware)
Coming from an internal API call, `req` is a `NextApiRequest` instance. (https://nextjs.org/docs/api-routes/introduction)

Reading headers needs to be handled differently here:
```
export async function getIdentityToken(req: NextApiRequest | NextRequest): Promise<JwtPayload> {
  let jwt = req.headers['x-forwarded-jwt'] ?? ''
  if (req instanceof NextRequest) {
    jwt = req.headers.get('x-forwarded-jwt')
  }
  return jwtDecode(String(jwt))
}
```

Is there a more streamlined or well-practiced way of handling this without having to check the object type?

**Comment by icyJoseph** [comment]
I would check for `headers instanceof Headers`, and use native classes and types. I used to do a check for `get`, but looking back at my notes, `instanceof Headers` ought to be more "correct".

```ts
import type { IncomingMessage } from "http";

export const hasMappedHeaders = (headers: Headers | IncomingMessage['headers']): headers is Headers => {
    return headers instanceof Headers;
};

// in another file...
export async function getIdentityToken(req: Request | IncomingMessage): Promise<JwtPayload> {
    const jwt = hasMappedHeaders(req.headers)
        ? req.headers.get('x-forwarded-jwt')
        : req.headers['x-forwarded-jwt'];

    return jwtDecode(String(jwt ?? '').replace('Bearer ', ''));
}
```

Never mind my terrible naming criteria. This way this function can work in any environment, `edge` or `node.js`, because types, like `IncomingMessage`, are stripped away, and `Headers` is a common object to both `edge` and `node.js`.

You could also wrap all of this into a function that takes either type of headers and a key to retrieve, but some might say that's a hasty abstraction.
